ui/src/shared/components/OptIn.js

```diff
@@ -0,0 +1,182 @@
+import React, {Component, PropTypes} from 'react'
+import classnames from 'classnames'
+
+import uuid from 'node-uuid'
+
+import ClickOutsideInput from 'shared/components/ClickOutsideInput'
+
+class OptIn extends Component {
+  constructor(props) {
+    super(props)
+
+    const {customValue, fixedValue} = props
+
+    this.state = {
+      useCustomValue: customValue !== '',
+      fixedValue,
+      customValue,
+    }
+
+    this.id = uuid.v4()
+    this.isCustomValueInputFocused = false
+
```

Comment:
I'm not really sure, but I don't think it should, actually. Since `setState` is async, and this flag is meant to be used in sync with UI events and to guard against double-clicks, it seemed safest to keep it outside of React so that it would be instantly available. It should not trigger a re-render, and nothing in render or downstream is consuming it, rather it's only for arbitrating click events.